WalletWasabi.Fluent/ViewModels/Wallets/CoinJoinStateViewModel.cs

```diff
@@ -0,0 +1,282 @@
+using System.Reactive;
+using System.Reactive.Linq;
+using System.Windows.Input;
+using Avalonia.Threading;
+using NBitcoin;
+using ReactiveUI;
+using WalletWasabi.Fluent.Helpers;
+using WalletWasabi.Fluent.State;
+using WalletWasabi.WabiSabi.Client;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.Fluent.ViewModels.Wallets;
+
+public partial class CoinJoinStateViewModel : ViewModelBase
+{
+	enum State
+	{
+		AutoCoinJoin,
+		ManualCoinJoin,
+
+		AutoStarting,
+		Paused,
+		AutoPlaying,
+
+		Stopped,
+		ManualPlaying,
+	}
+
+	enum Trigger
+	{
+		AutoCoinJoinOn,
+		AutoCoinJoinOff,
+		AutoCoinJoinEntered,
+		ManualCoinJoinEntered,
+		Pause,
+		Play,
+		Stop,
+		PlebStop,
+		RoundStartFailed,
+		RoundStart
+	}
+
+	private readonly StateMachine<State, Trigger> _machine;
+	private readonly Wallet _wallet;
+
+	[AutoNotify] private bool _isAutoWaiting;
+	[AutoNotify] private bool _isAuto;
+	[AutoNotify] private bool _playVisible = true;
+	[AutoNotify] private bool _pauseVisible;
+	[AutoNotify] private bool _stopVisible;
+	[AutoNotify] private MusicStatusMessageViewModel? _currentStatus;
+	[AutoNotify] private bool _isProgressReversed;
+	[AutoNotify] private double _progressValue;
+	[AutoNotify] private string _elapsedTime;
+	[AutoNotify] private string _remainingTime;
+
+	private readonly MusicStatusMessageViewModel _countDownMessage = new()
+		{ Message = "Waiting to auto-start coinjoin" };
+
+	private readonly MusicStatusMessageViewModel _coinJoiningMessage = new() { Message = "Coinjoining" };
+
+	private readonly MusicStatusMessageViewModel _pauseMessage = new()
+		{ Message = "Coinjoin is paused" };
+
+	private readonly MusicStatusMessageViewModel _stoppedMessage = new() { Message = "Coinjoin is stopped" };
+
+	private DateTimeOffset _autoStartTime;
+	private DateTimeOffset _countDownStarted;
+
+	public CoinJoinStateViewModel(WalletViewModel walletVm, IObservable<Unit> balanceChanged)
+	{
+		_elapsedTime = "";
+		_remainingTime = "";
+
+		_wallet = walletVm.Wallet;
+
+		var timer = new DispatcherTimer
+		{
+			Interval = TimeSpan.FromSeconds(1),
+		};
+
+		var coinJoinManager = Services.HostedServices.Get<CoinJoinManager>();
+
+		Observable.FromEventPattern<StatusChangedEventArgs>(coinJoinManager, nameof(coinJoinManager.StatusChanged))
```

Comment:
I'd add a `DisposeWith` to explicitly dispose subscriptions, although it might not be necessary according to [this](https://www.reactiveui.net/docs/guidelines/framework/dispose-your-subscriptions)

Reply:
will deal with this after release. needs a fairly big refactor. and these objects lifetime is basically until program closes so wont cause an issue.